modules/core/src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala

```diff
@@ -44,86 +44,205 @@ final case class Resolver[F[_]](repos: List[Registry], cache: Option[ResolverCac
+   * If any of repositories gives non-non-found error, lookup will retried
+   *
+   * @param schemaKey The SchemaKey uniquely identifying the schema in Iglu
+   * @return a Validation boxing either the Schema's
+   *         Json on Success, or an error String
+   *         on Failure
+   */
+  def lookupSchema(
+    schemaKey: SchemaKey
+  )(implicit
+    F: Monad[F],
+    L: RegistryLookup[F],
+    C: Clock[F]
+  ): F[Either[ResolutionError, Json]] =
+    lookupSchemaResult(schemaKey).map(_.map(_.value))
+
   /**
    * Get list of available schemas for particular vendor and name part
    * Server supposed to return them in proper order
    */
-  def listSchemas(
+  def listSchemasResult(
     vendor: Vendor,
     name: Name,
     model: Model
-  )(implicit F: Monad[F], L: RegistryLookup[F], C: Clock[F]) = {
+  )(implicit
+    F: Monad[F],
+    L: RegistryLookup[F],
+    C: Clock[F]
+  ): F[Either[ResolutionError, SchemaListLookupResult]] = {
     val get: Registry => F[Either[RegistryError, SchemaList]] = r => L.list(r, vendor, name, model)
 
-    val resultAction = cache
-      .fold(F.pure(none[ListLookup]))(_.getSchemaList(vendor, name, model))
-      .flatMap {
-        case Some(listResult) =>
-          listResult.fold(retryCached[F, SchemaList](get, vendor), finish[F, SchemaList])
+    def handleAfterFetch(
+      result: Either[LookupFailureMap, SchemaList]
+    ): F[Either[ResolutionError, SchemaListLookupResult]] =
+      cache match {
+        case Some(c) =>
+          c.putSchemaListResult(vendor, name, model, result).map {
+            case Right(ResolverCache.TimestampedItem(schemaList, timestamp)) =>
+              Right(ResolverResult.Cached((vendor, name, model), schemaList, timestamp))
+            case Left(failure) => Left(resolutionError(failure))
+          }
         case None =>
-          traverseRepos[F, SchemaList](get, prioritize(vendor, allRepos.toList), Map.empty)
+          result
+            .bimap[ResolutionError, SchemaListLookupResult](
+              resolutionError,
+              ResolverResult.NotCached(_)
+            )
+            .pure[F]
       }
 
-    for {
-      result <- resultAction
-      _      <- cache.traverse(c => c.putSchemaList(vendor, name, model, result))
-    } yield postProcess(result)
+    getSchemaListFromCache(vendor, name, model).flatMap {
+      case Some(TimestampedItem(Right(schemaList), timestamp)) =>
+        Monad[F].pure(Right(ResolverResult.Cached((vendor, name, model), schemaList, timestamp)))
+      case Some(TimestampedItem(Left(failures), _)) =>
+        retryCached[F, SchemaList](get, vendor)(failures)
+          .flatMap(handleAfterFetch)
+      case None =>
+        traverseRepos[F, SchemaList](get, prioritize(vendor, allRepos.toList), Map.empty)
+          .flatMap(handleAfterFetch)
+    }
   }
 
+  /**
+   * Get list of available schemas for particular vendor and name part
+   * Server supposed to return them in proper order
+   */
+  def listSchemas(
```

Comment:
`listSchemasResult` uses type aliases for vendor, name and model. Can we use it here as well?

Reply:
Absolutely. I think I've got them all now.